### ISSUE TITLE
Idea: a hack to ease the usage of toolbox

### ISSUE THREAD

**Issue by liufengyun** [liufengyun/gestalt#36]
Following the idea of declaring a dummy toolbox:

```Scala
package object gestalt {
    implicit val toolbox: Toolbox = null
}
```

To walk around the type system, it seems it's possible to reflectively set this immutable field to the values that are required in macro expansion. By doing this, we avoid issues with path-dependent types, and programmers always use the dummy toolbox.

Not sure if it's a feasible or good hack.

**Comment by liufengyun** [comment]
Could you share your thoughts on this @xeno-by ?

**Comment by liufengyun** [comment]
I do see the split of `Toolbox` into `StructToolbox` and `TypeToolbox` is in tension with this idea. If this idea works, I'm happy to merge the two boxes (maybe use implicit capabilities to restrict the API usage).

**Comment by xeno-by** [comment]
Haha that's an interesting idea!

In addition to importing `gestalt._`, users will also have to import `toolbox._`, but it's much better than having to deal with the usual path-dependent complications. 

The remaining inconvenience can be further reduced by making the package object extend `Toolbox` and then having it to delegate its methods to the global `toolbox` :)

**Comment by liufengyun** [comment]
Thanks a lot @xeno-by . I see another concern with this approach is thread safety, but the Dotty compiler assumes no concurrency, may be we can generalise the assumption to all Scala implementations.

**Comment by valdisxp1** [comment]
I have an alternative (full code at valdisxp1/gestalt@11abbe0e027b94b66756ab2492bf2ab18233acd8):
```scala
class HelperExample[T <: Toolbox](val tb: T) {
  def intArrayLit(array: Int*): tb.Tree = tb.Apply(tb.Ident("Array"), array.toList.map(tb.Lit))
}

class HelperExample2[T <: TypeToolbox](val tb: T) {
  def intArrayLit(array: Int*): tb.Tree = tb.Apply(tb.Ident("Array"), array.toList.map(tb.Lit))

  object Varargs {
    def unapply(tree: tb.Tree): Option[Either[tb.Tree, Seq[tb.Tree]]] = tree match {
      case tb.SeqLiteral(seq) => Some(Right(seq))
      case tb.Ascribe(seq, _) => Some(Left(seq))
      case _ => None
    }
  }
}
```
I believe the typing nice because it explains the semantics of helpers: `HelperExample` should work for any subtype of Toolbox - any implementation at all. `HelperExample` should work for any subtype of TypeToolbox - any implementation that supports def macro specific extractors.

```scala
//usage
meta {
  val helper = new HelperExample[toolbox.type](toolbox)
  import helper._
//...
}
```
Here we have to be explicit: I want a helper for my `toolbox`. But not any old `Toolbox`, but the exact `toolbox` I was given by the macro expansion.

**Comment by xeno-by** [comment]
@valdisxp1 Indeed, this approach works (see section 3.1.5 of my dissertation) in my dissertation.

A significant problem associated with it is verbosity. Over the years, dozens of users expressed dissatisfaction with the amount of ceremony to encode this approach.

Also, the solution involves a lot of path-dependent machinery, and as such is noticeably hard to understand for people without significant skill in Scala.

**Comment by liufengyun** [comment]
Thanks a lot for the creative thinking, you independently discovered the same solution as put forward in Eugene's [thesis](https://infoscience.epfl.ch/record/226166).

The bound on different kind of Toolbox is very nice, to some extent it justifies the verbosity. That is the last resort if we cannot do better.

**Comment by liufengyun** [comment]
BTW, during a talk with @sjrd, thread safety is indeed a concern. I was told it's possible to have multiple Scalac compiler instance run in parallel. Thus, my originally proposed idea is problematic.

**Comment by valdisxp1** [comment]
@liufengyun thanks for the link
@xeno-by definitely will be read more of your thesis. This looks fun.

 If we want to magic away some ceremony, we can do that during macro transformation similar to what we are doing with `this` -> `prefix`. I am considering defining helpers as implicit classes and then having transform populate the `T` parameter for all invocations.
```scala
implicit class HelperExample[T <: Toolbox](val tb: T) {
   //...
}
```
I guess there is a good reason why the Typer does not infer it already?
```scala
type T <: toolbox.type & helper.tb & scala.gestalt.Toolbox
```

All other solutions I can think of may reduce the ceremony in one place while greatly increasing it elsewhere.

It is sad that we need to hide path-dependent types as they are so cool. However, I can image the scenario: "I just wanted to write a simple macro that has nothing to do with these path-dependent types, but now I still need to deal with them".

**Comment by xeno-by** [comment]
@valdisxp1 What do you find so cool about path-dependent types in this scenario?

**Comment by valdisxp1** [comment]
I still need to write it and make sure that works but the idea is to have transform populate them.
From the point of view macro writer who uses the helper:
```scala
  import example.helpers._ // unavoidable you must tell what helpers to include
//...
meta {
  val array =  toolbox.intArrayLit(1, 3, 3, 7)
//...
}
```
Each call to the helper is implicit, so it gets expanded.
```scala
val array =  toolbox.intArrayLit(1, 3, 3, 7)
// expands to
val array = new HelperExample(toolbox).intArrayLit(1, 3, 3, 7)
// the transformer populates the path dependent type argument
val array = new HelperExample[toolbox.type](toolbox).intArrayLit(1, 3, 3, 7)
```
Therefore path-dependent types are hidden from the macro writer as well as the one who wrote the helper who just need to have a type argument.

**Comment by xeno-by** [comment]
@valdisxp1 I think you're underestimating the overhead associated with path-dependent types in situations like this.

Related complexity manifests itself in multiple ways, including: 1) necessity to carry toolboxes around everywhere you want to metaprogram, 2) path-dependent types in method signatures, 3) overhead in associated error messages, 4) requirement for the users to understand the details of how path-dependent types work. Individually, these things seem like little inconveniences, but together they add up to an annoying user experience.

In a way, this is an instance of the framework vs library problem. A framework (like scalac compiler API, scala.reflect or gestalt) forces you structure your code in a special way. A library (like scala.meta) you just use, utilizing any programming techniques you fancy. For example, in scala.reflect, you are forced to write an extension method for a Universe, where in scala.meta, you just write a normal method, and so on.

5+ years ago, when I just joined the scala.reflect project, I was very much in favor of a path-dependent approach. Additional static guarantees seemed awesome (e.g. with path-dependent types, it was impossible to mix artifacts from compile-time and runtime universes), and the overhead seemed miniscule. 

Over the years, based on community feedback, I radically changed my position on the issue. Extra guarantees didn't bring us anything practical, whereas the extra ceremony got old pretty quickly, creating a barrier to entry for newcomers and imposing a complexity tax on experts.

**Comment by xeno-by** [comment]
@liufengyun Can we turn the global `toolbox` val into a thread-local var under the covers?

**Comment by liufengyun** [comment]
@xeno-by That's a good idea. But unless Scala supports some annotation like `@stable`, to treat a method as stable path, it seems difficult to get thread-local var type correct.

**Comment by xeno-by** [comment]
The global val can provide a proxy-like custom implementation of the `Toolbox` interface. For every method call, this implementation can look up a thread-local variable set by the macro engine and delegate the call to it.

**Comment by liufengyun** [comment]
> The global val can provide a proxy-like custom implementation of the Toolbox interface.

Could you elaborate more in detail? It seems to be not easy to turn a `val tb: Toolbox = ...` into a proxy or method call, while still be able to use it as a prefix.

**Comment by liufengyun** [comment]
Ah, I seem to get your idea. You mean something like the follows?

```Scala
package object gestalt {
  type Tree
  type TypeTree

  private val toolbox = new ThreadLocal[Toolbox]

  def Lit(value: Any): Tree = toolbox.get.Lit(value)
}
```

**Comment by xeno-by** [comment]
Yes!

**Comment by liufengyun** [comment]
Implemented in #80
